src/pages/hayoungGam/Login/LoginForm/LoginForm.js

```diff
@@ -18,73 +19,25 @@ class LoginForm extends Component {
     this.props.history.push('/list-hayoung');
   };
 
-  validateId = () => {
-    const { id } = this.state;
-    let idStatus = false;
-    let error;
-    const emailRegex =
-      /^[\w+-]+(.[\w+-]+)*@([a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]*\.)+[a-zA-Z]{2,}$/g;
-
-    if (!emailRegex.test(id)) {
-      error = '올바른 이메일을 입력해 주세요.';
+  validateInput = () => {
+    const newPwState = validatePw(this.state);
+    if (newPwState.pwStatus) {
+      const newIdState = validateId(this.state);
+      this.setState({
+        ...this.state,
+        ...newIdState,
+        pwStatus: newPwState.pwStatus, // pwStatus에 true를 저장
+      });
     } else {
-      error = '';
-      idStatus = true;
-    }
-
-    this.setState({
-      error,
-      idStatus,
-    });
-  };
-
-  validatePw = () => {
-    const { pw } = this.state;
-    let pwStatus = false;
-    let error;
-    const lowerCaseRegex = /[a-z]/g;
-    const upperCaseRegex = /[A-Z]/g;
-    const digitRegex = /\d/g;
-    const specialCharcsRegex = /[ `!@#$%^&*()_+\-=[\]{};':"\\|,.<>/?~]/g;
-    const minLengthRegex = /^.{8,}$/g;
-
-    switch (false) {
-      case !(pw === ''):
-        error = '';
-        break;
-      case lowerCaseRegex.test(pw):
-        error = '비밀번호는 영어 소문자를 포함해야 합니다.';
-        break;
-      case upperCaseRegex.test(pw):
-        error = '비밀번호는 영어 대문자를 포함해야 합니다.';
-        break;
-      case specialCharcsRegex.test(pw):
-        error = '비밀번호는 특수 문자를 포함해야 합니다.';
-        break;
-      case digitRegex.test(pw):
-        error = '비밀번호는 숫자를 포함해야 합니다.';
-        break;
-      case minLengthRegex.test(pw):
-        error = '비밀번호는 8자리 이상입니다.';
-        break;
-      default:
-        error = '';
-        pwStatus = true;
-        break;
-    }
-
-    this.setState({
-      error,
-      pwStatus,
-    });
-
-    if (pwStatus) {
-      this.validateId();
+      this.setState({ ...this.state, ...newPwState });
     }
   };
 
   handleInput = event => {
-    this.setState({ [event.target.name]: event.target.value }, this.validatePw);
+    this.setState(
+      { [event.target.name]: event.target.value },
+      this.validateInput
+    );
```

Comment:
state에 저장된 id, pw값을 기반으로 validation을 동작시켜야하기 때문에
state 값을 즉각적으로 이용하기 위한 목적으로 setState의 두번째 인자에서
`validateInput`함수를 호출하고 있습니다.

하지만 `validateInput`함수 내에서도 idStatus, pwStatus 플래그를 변경시켜주기 위해
setState를 또 한번 사용하고 있기 때문에...

콜백에 콜백을 부르고 있는 점이 굉장히 걸리고 [setState안에서 또한번 setState를 사용하고 있는 점](https://stackoverflow.com/questions/63128083/is-it-okay-to-have-setstate-in-callback-of-setstate),
렌더링이 2번 발생하고 있는 점이 우려됩니다😢

---

**1. 현재 코드를 개선해보기 위해 시도해본 것1️⃣**

그래서 위의 스택오버플로우 답변과 [공식 문서](https://ko.reactjs.org/docs/react-component.html#setstate)를 참고하여
다음과 같이 콜백이 아닌 update 함수를 만들어, 그 안에서 `validateInput` 함수를 호출하려는 시도해보았습니다.

```jsx
handleInput = event => {
  this.setState(prevState => {
    prevState.id = event.target.value;
    prevState.pw = event.target.value;
    return this.validateInput();
  });
};
```

하지만 동작이 원하는 대로 움직이지 않았고,
아래와 같은 에러가 발생했습니다.
```
index.js:1 Warning: An update (setState, replaceState, or forceUpdate) was scheduled from inside an update function. Update functions should be pure, with zero side-effects. Consider using componentDidUpdate or a callback.
```
![image](https://user-images.githubusercontent.com/50080535/122322906-31d8c800-cf61-11eb-8af8-df334451b6f4.png)

위의 에러를 보면 콜백 또는 componentDidUpdate를 사용하라고 안내를 해주고 있었습니다.

**2. 현재 코드를 개선해보기 위해 시도해본 것2️⃣**

그래서, 아래와 같이 componentDidUpdate를 사용해본 결과

```jsx
componentDidUpdate() {
  this.validateInput();
}

handleInput = event => {
  this.setState({ [event.target.name]: event.target.value });
};
```

```
Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
```
![image](https://user-images.githubusercontent.com/50080535/122323134-9431c880-cf61-11eb-8f81-0cff38a8425f.png)

componentDidUpdate 함수 내에서 setState를 호출하면 안된다는 것을 알려주고 있는 것 같습니다..

---

결국은 시도해본 것 1️⃣의 에러메시지에서 안내해준 방법인
콜백을 사용하는 방법밖에 없는것일까요?

혹시 이 외에 어떤 좋은 아이디어가 있을까요?
그리고 제가 시도해본 방법중에서 코드 자체가 잘못 쓰여진 부분이 있는 것 같다면 지적해주시면 감사하겠습니다.

잘부탁드립니다.